compiler/rustc_parse/src/parser/expr.rs

```diff
@@ -1733,6 +1736,18 @@ impl<'a> Parser<'a> {
             self.sess.gated_spans.gate(sym::async_closure, span);
         }
 
+        // Disable recovery for closure body
+        self.last_closure_body = Some(decl_hi);
+
+        if self.token.kind == TokenKind::Semi && self.token_cursor.frame.delim == DelimToken::Paren
+        {
+            // It is likely that the closure body is a block but where the
+            // braces have been removed. We will recover and eat the next
+            // statements later in the parsing process.
+
+            return Ok(self.mk_expr_err(lo.to(body.span)));
```

Comment:
What we might want to do is make a closure here where it's body is an error expression placeholder, although that doesn't change anything in this PR, I would like us to be able to mark that closure somehow (either by changing it's variant or just using the err expression as a marker) to detect cases where `{|x| let y = x + 1; y}` is written. But we can make that change after the fact.

Could we add a test for the "ruby closure" case? I'm intrigued at what it looks like under _this_ PR.

Reply:
Oh that makes sense. I had not noticed the difference between "a closure whose body is an expr error" and "an expr error" in my early development. Thanks

Reply:
@estebank I added the ruby-style closure syntax ui tests you talked about. In order to make tests pass, I had to annotate them with the (arguably not helpful) current compiler output. Please ping me if this is not was you expected :)